File: implementations/quicksort/javascript/QuicksortUnstable.js
// Unstable Quicksort
function quicksortUnstable(arr) {
  quicksortUnstableHelper(arr, 0, arr.length - 1);
}

function quicksortUnstableHelper(arr, left, right) {
  if (left >= right) {
    return;
  }
  const pivotIndex = partitionUnstable(arr, left, right);
  quicksortUnstableHelper(arr, left, pivotIndex - 1);
  quicksortUnstableHelper(arr, pivotIndex + 1, right);
}

function partitionUnstable(arr, left, right) {
  const pivotValue = arr[right];
  let i = left;
  for (let j = left; j < right; j++) {
    if (arr[j].compareTo(pivotValue) < 0) {
      [arr[i], arr[j]] = [arr[j], arr[i]];
      i++;
    }
  }
  [arr[i], arr[right]] = [arr[right], arr[i]];
  return i;
}

File: implementations/quicksort/cpp/quicksort_unstable.cpp
// Unstable Quicksort
template <typename T>
void quicksort_unstable(T* arr, int left, int right) {
    if (left >= right) {
        return;
    }
    int pivot_index = partition_unstable(arr, left, right);
    quicksort_unstable(arr, left, pivot_index - 1);
    quicksort_unstable(arr, pivot_index + 1, right);
}

template <typename T>
int partition_unstable(T* arr, int left, int right) {
    T pivot_value = arr[right];
    int i = left;
    for (int j = left; j < right; j++) {
        if (arr[j] < pivot_value) {
            std::swap(arr[i], arr[j]);
            i++;
        }
    }
    std::swap(arr[i], arr[right]);
    return i;
}
